hphp/util/asm-x64-intelxed.h

```diff
@@ -0,0 +1,1031 @@
+    XED_EMIT_ENCODE();
+    XED_ASSERT_ERR("arg");
+    return encodedSize;
+  }
+
+  ALWAYS_INLINE
+  void xedEmit1(xed_iclass_enum_t instr, const xed_encoder_operand_t& op,
+                    xed_uint_t effOperandSizeBits = 0) {
+    codeBlock.moveFrontier(xedEmit1(instr, op, effOperandSizeBits, dest()));
+  }
+
+  ALWAYS_INLINE
+  uint32_t xedEmit2(xed_iclass_enum_t instr,
+                    const xed_encoder_operand_t& op_1,
+                    const xed_encoder_operand_t& op_2,
+                    xed_uint_t effOperandSizeBits,
+                    CodeAddress destination) {
+    XED_EMIT_BEGIN();
+    xed_inst2(&instruction, s_xedState, instr, effOperandSizeBits, op_1, op_2);
+    XED_EMIT_ENCODE();
+    XED_ASSERT_ERR("arg, arg");
+    return encodedSize;
+  }
+
+  ALWAYS_INLINE
+  void xedEmit2(xed_iclass_enum_t instr, const xed_encoder_operand_t& op_1,
+                const xed_encoder_operand_t& op_2,
+                xed_uint_t effOperandSizeBits = 0) {
+    codeBlock.moveFrontier(xedEmit2(instr, op_1, op_2, effOperandSizeBits,
+                           dest()));
+  }
+
+  ALWAYS_INLINE
+  uint32_t xedEmit3(xed_iclass_enum_t instr,
+                    const xed_encoder_operand_t& op_1,
+                    const xed_encoder_operand_t& op_2,
+                    const xed_encoder_operand_t& op_3,
+                    xed_uint_t effOperandSizeBits,
+                    CodeAddress destination) {
+    XED_EMIT_BEGIN();
+    xed_inst3(&instruction, s_xedState, instr, effOperandSizeBits,
+              op_1, op_2, op_3);
+    XED_EMIT_ENCODE();
+    XED_ASSERT_ERR("arg, arg, arg");
+    return encodedSize;
+  }
+
+  ALWAYS_INLINE
+  void xedEmit3(xed_iclass_enum_t instr,
+                const xed_encoder_operand_t& op_1,
+                const xed_encoder_operand_t& op_2,
+                const xed_encoder_operand_t& op_3,
+                xed_uint_t effOperandSizeBits = 0) {
+    codeBlock.moveFrontier(xedEmit3(instr, op_1, op_2, op_3, effOperandSizeBits,
+                           dest()));
+  }
+
+public:
+  void emitInt3s(int n) {
+    if (n == 0) return;
+    static uint8_t instr = 0;
+    static uint32_t int3size = xedEmit0(XED_ICLASS_INT3, sz::byte, &instr);
+    static_cast<void> (int3size);
+    assert(int3size == 1);
+    for (auto i = 0; i < n; ++i) {
+      byte(instr);
+    }
+  }
+
+  void emitNop(int n) {
+    if (n == 0) return;
+    static const xed_iclass_enum_t nops[] = {
+      XED_ICLASS_INVALID,
+      XED_ICLASS_NOP,
+      XED_ICLASS_NOP2,
+      XED_ICLASS_NOP3,
+      XED_ICLASS_NOP4,
+      XED_ICLASS_NOP5,
+      XED_ICLASS_NOP6,
+      XED_ICLASS_NOP7,
+      XED_ICLASS_NOP8,
+      XED_ICLASS_NOP9,
+    };
+    // While n >= 9, emit 9 byte NOPs
+    while (n >= 9) {
+      xedInstr(XED_ICLASS_NOP9, 0);
+      n -= 9;
+    }
+    // Emit remaining NOPs (if any)
+    if (n) {
+      xedInstr(nops[n], 0);
+    }
+  }
+
+  void pad() {
+    auto remaining = available();
+    static uint32_t int3 = 0;
```

Comment:
You're emitting to a `uint32_t` rather than `uint8_t[XED_MAX_INSTRUCTION_BYTES]` because these instructions both fit within 4 bytes, right? That's hardcoding some information about the size of them, so at that point you might as well just do this instead:
```c++
static uint8_t int3[1];
static uint8_t ud2[2]; // or whatever their sizes actually are
```
That should also let you get rid of the casts on 722 and 726.